### PR TITLE
KAFKA-7890: Invalidate ClusterConnectionState cache for a broker if the hostname of the broker changes.

### PR DESCRIPTION
**Pull request by markcho**
See https://issues.apache.org/jira/browse/KAFKA-7890 for additional details.

- Added a new unit test where the hostname changes.
- Also tested in production

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by cmccabe** [comment]
This looks reasonable.  Can you log a message when the hostname changes?  It should be very infrequent.

**Comment by markcho** [comment]
Thanks for taking a look @cmccabe.

Added log.info for broker hostname changing.

File: clients/src/main/java/org/apache/kafka/clients/ClusterConnectionStates.java
Comment (rajinisivaram) [inline clients/src/main/java/org/apache/kafka/clients/ClusterConnectionStates.java]: nit: We don't use the 'get` prefix for getters, just `host()` to be consistent.

File: clients/src/test/java/org/apache/kafka/clients/ClusterConnectionStatesTest.java
Comment (rajinisivaram) [inline clients/src/test/java/org/apache/kafka/clients/ClusterConnectionStatesTest.java]: I can see why this test was removed, but I think there is a missing test case now. We are testing host name change with the updated test `testNodeWithNewHostname` that verifies the change from this PR. But we don't have a test that verifies the additional lookup added under KAFKA-7755 which was being tested earlier with a hostname change? We need a test that fails if we removed the look up added under KAFKA-7755.

**Review by rajinisivaram — COMMENTED** [review]
@markcho Thanks for the PR, looks good. Left a couple of minor comments.

File: clients/src/main/java/org/apache/kafka/clients/ClusterConnectionStates.java
Comment (markcho) [reply clients/src/main/java/org/apache/kafka/clients/ClusterConnectionStates.java]: Done

File: clients/src/test/java/org/apache/kafka/clients/ClusterConnectionStatesTest.java
Comment (markcho) [reply clients/src/test/java/org/apache/kafka/clients/ClusterConnectionStatesTest.java]: I restored the existing test and just updated a single line to work with the new code.

**Comment by rajinisivaram** [comment]
retest this please

**Review by rajinisivaram — APPROVED** [review]
@markcho Thanks for the updates, LGTM.

The test failures don't look related, but retrying the build once to see if we can get a clean build before merging. Will merge after builds complete.

**Comment by rajinisivaram** [comment]
Test failure not related. Merging to trunk, 2.2 and 2.1.

**Pull request merged**
